sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala

```diff
@@ -183,34 +182,131 @@ private[sql] object ArrowConverters {
   }
 
   /**
-   * Convert a byte array to an ArrowRecordBatch.
+   * Load a serialized ArrowRecordBatch.
    */
-  private[arrow] def byteArrayToBatch(
+  private[arrow] def loadBatch(
       batchBytes: Array[Byte],
       allocator: BufferAllocator): ArrowRecordBatch = {
-    val in = new ByteArrayReadableSeekableByteChannel(batchBytes)
-    val reader = new ArrowFileReader(in, allocator)
-
-    // Read a batch from a byte stream, ensure the reader is closed
-    Utils.tryWithSafeFinally {
-      val root = reader.getVectorSchemaRoot  // throws IOException
-      val unloader = new VectorUnloader(root)
-      reader.loadNextBatch()  // throws IOException
-      unloader.getRecordBatch
-    } {
-      reader.close()
-    }
+    val in = new ByteArrayInputStream(batchBytes)
+    MessageSerializer.deserializeMessageBatch(new ReadChannel(Channels.newChannel(in)), allocator)
+      .asInstanceOf[ArrowRecordBatch]  // throws IOException
   }
 
+  /**
+   * Create a DataFrame from a JavaRDD of serialized ArrowRecordBatches.
+   */
   private[sql] def toDataFrame(
-      payloadRDD: JavaRDD[Array[Byte]],
+      arrowBatchRDD: JavaRDD[Array[Byte]],
       schemaString: String,
       sqlContext: SQLContext): DataFrame = {
-    val rdd = payloadRDD.rdd.mapPartitions { iter =>
+    val schema = DataType.fromJson(schemaString).asInstanceOf[StructType]
+    val timeZoneId = sqlContext.sessionState.conf.sessionLocalTimeZone
+    val rdd = arrowBatchRDD.rdd.mapPartitions { iter =>
       val context = TaskContext.get()
-      ArrowConverters.fromPayloadIterator(iter.map(new ArrowPayload(_)), context)
+      ArrowConverters.fromBatchIterator(iter, schema, timeZoneId, context)
     }
-    val schema = DataType.fromJson(schemaString).asInstanceOf[StructType]
     sqlContext.internalCreateDataFrame(rdd, schema)
   }
+
+  /**
+   * Read a file as an Arrow stream and return an RDD of serialized ArrowRecordBatches.
+   */
+  private[sql] def readArrowStreamFromFile(sqlContext: SQLContext, filename: String):
+  JavaRDD[Array[Byte]] = {
+    val fileStream = new FileInputStream(filename)
+    try {
+      // Create array so that we can safely close the file
+      val batches = getBatchesFromStream(fileStream.getChannel).toArray
+      // Parallelize the record batches to create an RDD
+      JavaRDD.fromRDD(sqlContext.sparkContext.parallelize(batches, batches.length))
+    } finally {
+      fileStream.close()
+    }
+  }
+
+  /**
+   * Read an Arrow stream input and return an iterator of serialized ArrowRecordBatches.
+   */
+  private[sql] def getBatchesFromStream(in: SeekableByteChannel): Iterator[Array[Byte]] = {
+
+    // TODO: simplify in super class
+    class RecordBatchMessageReader(inputChannel: SeekableByteChannel) {
```

Comment:
I had to modify the existing Arrow code to allow for this, but I will work on getting these changes into Arrow for 0.10.0 and then this class can be simplified a lot.

Reply:
made https://issues.apache.org/jira/browse/ARROW-2704 to track

Reply:
I wonder if this works?
In a for-loop:
(1) Read next batch into VectorSchemaRoot (copy into arrow memory)
(2) Use VectorUnloader to unload the VectorSchemaRoot to an ArrowRecordBatch (no copy)
(3) Use MessageSerializer.serialize to write ArrowRecordBatch to a ByteChannel (copy from arrow memory to java memory)

Seems that we cannot directly read from socket into java memory anyway (have to go through Arrow memory allocator)..

Reply:
@BryanCutler Now I looked at this more I think I understand what are you trying to do: The whole class here is trying to read Arrow record batches from an stream into Java on-heap memory without going through Arrow off-heap memory, is that correct?

Also, this function is only used for pandas DataFrame -> Spark DataFrame?

Reply:
>The whole class here is trying to read Arrow record batches from an stream into Java on-heap memory without going through Arrow off-heap memory, is that correct?

Yes, that's correct.  This is done to parallelize the Arrow record batches.

> Also, this function is only used for pandas DataFrame -> Spark DataFrame?

Yes, `RecordBatchMessageReader` is a specialized class and only meant for this purpose.

Reply:
I see. Since this codepath already involves multiple copy of data (python -> disk -> jvm), I am not sure reducing one memory copy is worth the complexity. I feel at least we should have this change in the separate PR maybe? What do you think?

Reply:
I think we should definitely avoid extra copies of the data in the JVM if possible, since we are trying to be efficient here.  This process doesn't really seem complex to me, the specialized code here is about 10 lines and just reads bytes from an input stream to a byte buffer.  Can you clarify a bit more on why you think this should be in a separate PR?

Reply:
Sure. The main reason that I am not sure about this code is that the code here breaks encapsulation.

If I understand correctly, Arrow reader only supports reading record batch from Channel to Arrow memory, in order to read record batch from Channel to on-heap memory directly, we need to subclass `MessageChannelReader`, overwrite `readNextMessage` to load the the metadata and body of record batch.

Now the main point that I feel not comfortable with this approach:
The subclass changes the behavior of `readNextMessage` to load  both metadata and body of a record batch, where in the parent class it only loads meta of a record batch. And I think this is the contract of the interface too so this feels a bit hacky.

I am not saying I am totally against this for performance reasons, but considering the code path already involves writing data to disk (so avoid one memory copy won't necessary get us much) and is one of the less frequent operations (pandas DataFrame -> spark DataFrame), I am not sure it's worth it, that's why I suggest resolving this separately so not to block this PR.

Reply:
Btw I don't mind discuss it in this PR too. Also curious what other people think.

Reply:
Ok...  I don't really need to subclass `MessageChannelReader` here.  What if instead we just make a couple of static functions on the Arrow side to help with the details of processing messages, like:

```Java
public class MessageChannelReader {
...

  public static Integer readMessageLength(ReadChannel in) {..}

  public static Message loadMessage(ReadChannel in, int messageLength, ByteBuffer buffer) {..}
}
```
Is that better?

Reply:
Yeah.. one way to do it is to write a new MessageReader interface to read Arrow message from a Channel:

```
public class OnHeapMessageChannelReader {
  /**
   * Read the next message in the sequence.
   *
   * @return The read message or null if reached the end of the message sequence
   * @throws IOException
   */
  Message readNextMessage() throws IOException;

  /**
   * When a message is followed by a body of data, read that data into an ArrowBuf. This should
   * only be called when a Message has a body length > 0.
   *
   * @param message Read message that is followed by a body of data
   * @param allocator BufferAllocator to allocate memory for body data
   * @return An ArrowBuf containing the body of the message that was read
   * @throws IOException
   */
   ByteBuffer readMessageBody(Message message) throws IOException;

   ...
}
```

We might need to duplicate some logic in https://github.com/apache/arrow/blob/master/java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageChannelReader.java#L33 

For record batches, it's not too bad because the logic is pretty simple, but the down side is we will be using low level APIs of Arrow, which might not be guaranteed to be stable .

@BryanCutler what kind of static function do you think we need to add on the Arrow side?

Reply:
I was just referring to the two static functions in the previous post.  These will contain most of the low level operations inside for stability.  I'm not sure we need a new interface to handle this case, it's probably not a common use case.  I'll just implement what I thought and maybe it will be more clear.